### ISSUE TITLE
Adding a parameter to locally control number of threads

### ISSUE THREAD

**Issue by traversc** [RcppCore/RcppParallel#122]
I think this would be quite helpful for both users (as after compilation with `sourceCpp`, `RcppParallel::setThreadOptions` has no effect) and package makers (to offer more fine tuned control). 

In TBB, this can be done using a `task_arena` (https://stackoverflow.com/a/25435359/2723734).

Example:

```
inline void parallelFor2(std::size_t begin, std::size_t end, Worker& worker, 
                                     std::size_t grainSize = 1, std::size_t nthreads = 1) {
  tbb::task_arena limited(nthreads);
  tbb::task_group tg;
  limited.execute([&]{
    tg.run([&]{
      parallelFor(begin, end, worker, grainSize);
    });
  });
  limited.execute([&]{ tg.wait(); });
}
```

Full minimal example below. 

```
y <- psqrt2(1:1e6, nthreads=4) # psqrt2 prints "T" once per thread
TTTT
y <- psqrt2(1:1e6, nthreads=100)
TTTTTTTT
```

Rcpp script:
```
#include <Rcpp.h>
#include <RcppParallel.h>
#include "tbb/concurrent_unordered_map.h"
#include "tbb/task_arena.h"
using namespace Rcpp;
using namespace RcppParallel;

// [[Rcpp::depends(RcppParallel)]]
// [[Rcpp::plugins(cpp11)]]

inline void parallelFor2(std::size_t begin, std::size_t end, Worker& worker, 
                                     std::size_t grainSize = 1, std::size_t nthreads = 1) {
  tbb::task_arena limited(nthreads);
  tbb::task_group tg;
  limited.execute([&]{
    tg.run([&]{
      parallelFor(begin, end, worker, grainSize);
    });
  });
  limited.execute([&]{ tg.wait(); });
}

struct SquareRoot : public Worker {
  tbb::enumerable_thread_specific<bool> printme;
  const RVector<double> input;
  RVector<double> output;
  SquareRoot(const NumericVector input, NumericVector output) 
    : printme(false), input(input), output(output) {}
  
  void operator()(std::size_t begin, std::size_t end) {
    tbb::enumerable_thread_specific<bool>::reference printme_local = printme.local();
    if(!printme_local) { // print once per thread
      std::cout << "T";
      printme_local = true;
    }
    std::transform(input.begin() + begin, 
                   input.begin() + end, 
                   output.begin() + begin, 
                   ::sqrt);
  }
};

// [[Rcpp::export(rng = false)]]
NumericVector psqrt2(NumericVector x, int nthreads) {
  NumericVector output(Rf_xlength(x));
  SquareRoot squareRoot(x, output);
  parallelFor2(0, Rf_xlength(x), squareRoot, 1, nthreads);
  std::cout << std::endl;
  return output;
}
```

**Comment by kevinushey** [comment]
Thanks! This is a great suggestion. I've committed this here:

https://github.com/RcppCore/RcppParallel/commit/20353e34b0ace5207e23021bb35f7bc4ea3a9bee

(note the use of functors rather than lambdas just because I want to maintain compatibility with C++98, at least for now -- we can probably relax this next release)

**Comment by traversc** [comment]
Looks good, if I could add a few more nitpicky comments:

The constructor for `task_arena` uses `int` (I mistakenly assumed it used `std::size_t`) it'd be more consistent to pass in an int.  

Checking for reasonable `numThreads` values. When I gave it a large number of threads, R sometimes crashed. 

E.g.:

```
inline void tbbParallelFor(std::size_t begin,
                           std::size_t end, 
                           Worker& worker,
                           std::size_t grainSize = 1,
                           int numThreads = -1)
{
   int max_threads = tbb::task_scheduler_init::default_num_threads();
   tbb::task_arena arena(numThreads < 0 ? tbb::task_arena::automatic : std::min(numThreads, max_threads));

....
```

**Comment by kevinushey** [comment]
`tbb::task_scheduler_init::default_num_threads()` might be set to a lower value than one might want for a particular execution, though. Is the error specifically when one attempts to request more threads than that? Or something else?

**Comment by traversc** [comment]
I may be mistaken sorry, I can't reproduce the crash I observed last time I thought occurred due a large number of threads.

**Comment by kevinushey** [comment]
This has now been implemented in RcppParallel 5.1.2.
